### ISSUE TITLE
Request for clarification of change made in 2.9.7

### ISSUE THREAD

**Issue by jimlloyd** [petkaantonov/bluebird#487]
Hello. I have code using promisify that has worked without issues since at least version 2.3.11, through 2.9.6. With 2.9.7 the code fails.

I want to be able pass `require('bluebird').promisify` to a library that will create promisified methods of standard node async methods (unbound functions that depend on *this*). FYI, the library in question is [node-java](https://github.com/joeferner/node-java), with the feature [documented here](https://github.com/joeferner/node-java#promises). The feature was written to be workable with multiple Promises/A+ libraries. See the source file [testHelpers.js](https://github.com/joeferner/node-java/blob/master/testHelpers.js) to see (in comments) the various libraries that were tested. Most of the libraries (including bluebird) allowed me to pass a single promisify function and get valid results. Q was an exception, since the closest candidate function (denodify/nfbind) could only promisify a bound function, so I had to write a wrapper function promisifyQ show in testHelpers.js. It looks like that may now be necessary with bluebird.

Since this behavior change is recent, I want to ask for clarification. I see that Issue #470 may be the reason for the change in behavior, though that was specifically about promisifyAll, not promisify.

**Comment by petkaantonov** [comment]
The change should not affect any existing code at all.

**Comment by jimlloyd** [comment]
The change did affect my existing code, but my code arguably misused the API. The API documentation for Promise.promisify() clearly states:

*Note that if the node function is a method of some object, you need to pass the object as the second argument ...*

Despite this statement, I was able to pass unbound methods to promisify and get a valid promisified method. So, one way to look at this is that there has been a long-outstanding bug in which the implementation did not conform to the spec, and 2.9.7 fixes the bug. Another way to look at it is that the documentation was overly restrictive, and 2.9.7 inadvertently introduced an unnecessary change in a useful feature. Which way do you see it?

**Comment by petkaantonov** [comment]
If you don't pass a second argument, the promisified method uses whatever `this` is to call the original function. There is no change to that in 2.9.7 and that still works the same.

**Comment by jimlloyd** [comment]
Ok, it seems that I need to try to create a small test that reproduces the error I am now seeing. I see that your unit test 'when calling promisified function it should >  should use this if no receiver was given' that seems to be a good test for this, but there are some complexities to my particular case, including the fact that the promisification function is passed down into a node addon, so the actual call to promisify a method occurs in C++ code (lines 71-81 of https://github.com/joeferner/node-java/blob/master/src/javaObject.cpp).

But before doing that, I am trying to run your unit tests on my MacBook with Mac OS 10.10.2 using node 0.11.14. I'm not getting a clean run for the tests, and in fact I get this error:

Some tests failed: 
    promisify.js For additional details you can run it individually  with the command `node tools/test --run=promisify.js`

However, if I then run just promisify.js as suggested, it runs without errors. This behavior is repeatable. Every time I run all tests it fails, saying that promisify.js had the failures, but every time I run just promisfy.js, it succeeds (where 'every time' means ~5 times).

Perhaps the problem is somehow specific to my configuration. Do you test on a Mac?

FYI, when running all tests, I see many of these warnings:

```
Warning: .then() only accepts functions but was passed: [object Null], [object Array]
    at /Users/jim/dev/github/bluebird/test/mocha/3.2.6.js:305:45
    at Context.<anonymous> (/Users/jim/dev/github/bluebird/test/mocha/helpers/testThreeCases.js:59:23)
    at Test.Runnable.run (/Users/jim/dev/github/bluebird/node_modules/mocha/lib/runnable.js:218:15)
    at Runner.runTest (/Users/jim/dev/github/bluebird/node_modules/mocha/lib/runner.js:374:10)
    at /Users/jim/dev/github/bluebird/node_modules/mocha/lib/runner.js:452:12
    at next (/Users/jim/dev/github/bluebird/node_modules/mocha/lib/runner.js:299:14)
    at /Users/jim/dev/github/bluebird/node_modules/mocha/lib/runner.js:309:7
    at next (/Users/jim/dev/github/bluebird/node_modules/mocha/lib/runner.js:248:23)
    at Immediate._onImmediate (/Users/jim/dev/github/bluebird/node_modules/mocha/lib/runner.js:276:5)
    at processImmediate [as _immediateCallback] (timers.js:374:17)
```

And then these two errors:

```
Possibly unhandled AssertionError: false == true
    at assertLongStackTraces (/Users/jim/dev/github/bluebird/test/mocha/promisify.js:550:5)
    at /Users/jim/dev/github/bluebird/test/mocha/promisify.js:566:21
    at process._tickCallback (node.js:372:11)
Possibly unhandled AssertionError: false == true
    at assertLongStackTraces (/Users/jim/dev/github/bluebird/test/mocha/promisify.js:550:5)
    at /Users/jim/dev/github/bluebird/test/mocha/promisify.js:584:21
    at process._tickCallback (node.js:372:11)
```

**Comment by petkaantonov** [comment]
The warning is ok and expected as it's the Promises/A+ test suite testing behavior when  `.then` is incorrectly called with non function values. In practice this warning helps people new to promises quickly see where they made a mistake.

I used 0.11.14 and it fails on ubuntu too. 0.11.16 is fine, so is 0.12 and iojs. But even on 0.11.14 that failure doesn't really affect anything other than long stack traces.

**Comment by jimlloyd** [comment]
Interesting. I can confirm that your unit tests all pass when I use node 0.12 on my MacBook. Yet my test case in node-java is still failing. I'll investigate more.

**Comment by petkaantonov** [comment]
@jimlloyd what is your java-test that fails? Can you also log out the generated code of the promisified function e.g.:

```js
var promisifiedFunction = Promise.promisify(someFunction);
var generatedSourceCode = promisifiedFunction.toString();
console.log(generatedSourceCode);
```

**Comment by jimlloyd** [comment]
I'm currently debugging in some non-public code. Here is the generated code for my failing case:

```javascript
function (_arg0,_arg1,_arg2) {                                    
    'use strict';                                                    
    var len = arguments.length;                                      
    var promise = new Promise(INTERNAL);                             
    promise._captureStackTrace();                                    
    var nodeback = nodebackForPromise(promise);                      
    var ret;                                                         
    var callback = tryCatch(fn);                                     
    switch(len) {                                                    
        case 0:ret = callback.call(receiver, nodeback); break;
        case 1:ret = callback.call(receiver, _arg0, nodeback); break;
        case 2:ret = callback.call(receiver, _arg0, _arg1, nodeback); break;
        case 3:ret = callback.call(receiver, _arg0, _arg1, _arg2, nodeback); break;
                                                 
        default:                                                             
            var args = new Array(len + 1);                                   
            var i = 0;                                                       
            for (var i = 0; i < len; ++i) {                                  
               args[i] = arguments[i];                                       
            }                                                                
            args[i] = nodeback;                                              
            ret = callback.apply(receiver, args);
            break;                                                           
    }                                                                
    if (ret === errorObj) {                                          
        promise._rejectCallback(maybeWrapAsError(ret.e), true, true);
    }                                                                
    return promise;                                                  
}
```

I assume several of the variables used here are closure variables defined outside the scope of this generated function, in particular `receiver`. 

I'm still debugging, but have made slow progress today as I have had many distractions. My use case is a strange one, involving not only node addons, but an addon that brings up an embedded JVM and calls java functions. So, my use case is very much a corner case. If you don't see anything weird in the above, then I should do more work to pin down the bug on my end and get back to you. Thanks.

**Comment by petkaantonov** [comment]
Since the generated code has  `.call(receiver)` rather than `.call(this)` you must have called `promisify` with a non-undefined second argument.

**Comment by jimlloyd** [comment]
See https://github.com/jimlloyd/bluebird-promises-corner-case

This still has the complexity that depend on node-java, but it is the best short illustration of the problem I have so far.

**Comment by petkaantonov** [comment]
Cannot make npm release right now but you can test with master by cloning

    git clone https://github.com/petkaantonov/bluebird.git
    cd bluebird && npm install
    node tools/build --main
    // generated files are in js/main folder

**Comment by jimlloyd** [comment]
Thanks! I was busy this morning but finally had time to test, and saw that you released 2.9.8. I just tested it and it works great.

I wonder if you should clarify the API documentation for promisify. I'll submit a PR with a suggested change to API.md.

**Comment by petkaantonov** [comment]
Yes, pull requests are very welcome :) thank you
